nomad/job_endpoint.go

```diff
@@ -298,6 +298,24 @@ func (j *Job) Register(args *structs.JobRegisterRequest, reply *structs.JobRegis
 			}
 		}
 	}
+	// Preserve the existing task resources, if requested
+	if existingJob != nil && args.PreserveResources {
+		prevResources := make(map[string]map[string]*structs.Resources)
+		for _, tg := range existingJob.TaskGroups {
+			prevResources[tg.Name] = make(map[string]*structs.Resources)
+			for _, task := range tg.Tasks {
+				prevResources[tg.Name][task.Name] = task.Resources
+			}
+
+		}
+		for _, tg := range args.Job.TaskGroups {
+			for _, task := range tg.Tasks {
+				if res, ok := prevResources[tg.Name][task.Name]; ok {
+					task.Resources = res
+				}
+			}
+		}
+	}
```

Comment:
I realize we're copying the existing logic for `PreserveCounts` here, but unfortunately I think that logic is racy with concurrent RPCs. Ex. supposed I have a `Job.Register` and `Job.Scale` RPC in-flight at the same time:

* `Job.Register` reads the resources
* `Job.Scale` updates the resources and writes to the FSM
* `Job.Register` writes to the FSM

So I think we need to move _both_ of these features into the state store code.